Lab-2/BinaryTree.cpp

```diff
@@ -0,0 +1,488 @@
+    return m_root->getRight();
+}
+BinaryTree BinaryTree::copyLeftChild() const{
+    return m_root->getLeft();
+}
+
+int BinaryTree::height() const {
+    return height(m_root);
+}
+
+int BinaryTree::height(int key) const
+{
+    if (find(key).isEmpty())
+        return -1;
+    return height(find(key).m_root);
+}
+
+int BinaryTree::height(Node* root) const {
+    if (root == nullptr)
+    {
+        return 0;
+    }
+    if (root->getRight() == nullptr && root->getLeft() == nullptr) {
+        return 1;
+    }
+    if (root->getRight()==nullptr) {
+        return height(root->getLeft())+1;
+    }
+    if (root->getLeft()==nullptr) {
+        return height(root->getRight())+1;
+    }
+    return std::max(height(root->getRight()), height(root->getLeft()))+1;
+}
+
+int BinaryTree::power() const {
+    return power(m_root);
+}
+int BinaryTree::power(Node* root) const {
+    if (root == nullptr)
+    {
+        return 0;
+    }
+    if (root->getRight() == nullptr && root->getLeft() == nullptr) {
+        return 1;
+    }
+    if (root->getRight()==nullptr) {
+        return power(root->getLeft())+1;
+    }
+    if (root->getLeft()==nullptr) {
+        return power(root->getRight())+1;
+    }
+    return power(root->getRight())+power(root->getLeft())+1;
+}
+
+int BinaryTree::max() const {
+    return max(m_root);
+}
+int BinaryTree::max(Node* root) const {
+    if (root->getRight() == nullptr && root->getLeft() == nullptr) {
+        return root->getKey();
+    }
+    if (root->getRight()==nullptr) {
+        return std::max(max(root->getLeft()), root->getKey());
+    }
+    if (root->getLeft()==nullptr) {
+        return std::max(max(root->getRight()), root->getKey());
+    }
+    return std::max(std::max(max(root->getRight()), max(root->getLeft())), root->getKey());
+}
+
+int BinaryTree::min() const {
+    return min(m_root);
+}
+int BinaryTree::min(Node* root) const {
+    if (root->getRight() == nullptr && root->getLeft() == nullptr) {
+        return root->getKey();
+    }
+    if (root->getRight()==nullptr) {
+        return std::min(min(root->getLeft()), root->getKey());
+    }
+    if (root->getLeft()==nullptr) {
+        return std::min(min(root->getRight()), root->getKey());
+    }
+    return std::min(std::min(min(root->getRight()), min(root->getLeft())), root->getKey());
+}
+
+BinaryTree BinaryTree::find(const int key) const
+{
+    BinaryTree temp = clone(_find(key, m_root));
+    return temp;
+}
+BinaryTree::Node* BinaryTree::_find(const int key, Node* root) const {
+    if (root==nullptr) {
+        return nullptr;
+    }
+    if (root->getKey()==key) {
+        return root;
+    }
+    else {
+        Node* searchRight = _find(key, root->getRight());
+        Node* searchLeft = _find(key, root->getLeft());
+        if (searchRight!=nullptr) {
+            return searchRight;
+        }
+        else if (searchLeft!=nullptr) {
+            return searchLeft;
+        }
+        else
+            return nullptr;
```

Comment:
Зачем сразу искать в левом поддереве, если мы уже могли найти узел в правом?
```suggestion
        Node* node = _find(key, root->getRight());
        if (node == nullptr) {
            node = _find(key, root->getLeft());
        }
        
        return node;
```